token/wallet.tact

```diff
@@ -2,164 +2,116 @@
+        .send(
+            ihr_disabled: true,
+            can_be_bounced: true,
+            carry_message_input: true,
+        );
     }
 
-    if (!response_address.is_null() && msg_value > 0) {
-        RequestBuilder.new()
-            .ihr_disabled()
-            .ignore_action_phase_errors()
-            .money(msg_value)
-            .body(Excesses {
-                query_id: msg.query_id,
-            })
-            .send_to(response_address);
+    internal receive_tokens#178d4519(transfer: InternalTransfer) {
+        verify(
+            msg.sender === self.master || 
+                sender == WalletAddress.from_data(msg.from, self.master, self.wallet_code), 
+            "sender must be master or user wallet"
+        );
+
+        let ton_balance_before_msg = self.my_balance() - msg.amount;
+        let storage_fee = STORAGE_MIN_CONSUMPTION - ton_balance_before_msg.min(STORAGE_MIN_CONSUMPTION);
+        let mut msg_value = msg.amount - (storage_fee + GAS_CONSUMPTION);
+
+        if (transfer.forward_ton_amount != 0) {
+            msg_value -= (transfer.forward_ton_amount + fwd_fee);
+            let payload = msg.forward_payload;
+            self.owner
+                .notify_transfer(
+                    transfer.query_id,
+                    transfer.amount,
+                    msg.from,
+                    payload,
+                )
+                .send(ihr_disabled: true, pay_transfer_fee_separately: true);
+        }
+
+        if (!transfer.response_address.is_none() && msg_value > 0) {
+            transfer
+                .response_address
+                .excesses(msg.query_id)
+                .send(
+                    money: msg.amount,
+                    ihr_disabled: true,
+                    ignore_action_phase_errors: true
+                );
+        }
     }
-}
 
-fn burn_tokens(
-    @mut state: WalletState, 
-    msg: Burn, 
-    sender: ContractAddress, 
-    msg_value: Coins
-    fwd_fee: Coins, 
-) {
-    verify(state.owner == sender, "sender must be owner");
-    verify(state.balance - msg.amount >= 0, "not enough balance");
-
-    let required_input = fwd_fee + 2 * GAS_CONSUMPTION;
-    verify(msg_value > required_input, "not enough value");
-
-    state.balance -= msg.amount;
-
-    RequestBuilder.new()
-        .ihr_disabled()
-        .can_be_bounced()
-        .with_coins(CarryMessageInput{})
-        .body_as_ref[InternalMsgBody](BurnNotification {
-            query_id: msg.query_id,
-            amount: msg.amount,
-            from: state.owner,
-            response_address: msg.response_destination,
-        })
-        .send_to(state.master);
-}
-
-fn on_bounce(
-    @mut state: WalletState,
-    body: Slice
-) {
-    let op = body.load_uint(32);
-    verify(op == InternalTransfer || op == BurnNotification, "invalid bounce operation");
+    internal burn_tokens#595f07bc(query_id: Uint64, burn_amount: JettonAmount, response_destination: ContractAddress) {
```

Comment:
```suggestion
    internal burn_tokens#595f07bc(query_id: Uint64, burn_amount: JettonAmount, response_destination: Address) {
```

Let's strive for keeping simpler terms for base features and composite terms for user-defined types.

`Contract`, `Address`, `Interface`.

Reply:
Address can be external or internal, so `Address` should be "any address" while `ContractAddress` is internal address.

Reply:
Fixed in [0f79cbf](https://github.com/tact-lang/tact-bikeshed/pull/5/commits/0f79cbfa559bb91401ce9e4af0f66ca232836751)